### ISSUE TITLE
git pull --rebase should be removed asap

### ISSUE THREAD

**Issue by rodrigopinto** [Helabs/jumpup#67]
Guys, `git pull --rebase` isn't a best pratice and can generate merge conflicts during integration. It doesn't make sense still to use this command. Pull & Merge must be made manually before run the integration.

**Comment by thiagogabriel** [comment]
I agree. `git pull --rebase` almost always mess up my merges as well as @mikaelcarrara's.
Are you suggesting `git pull` instead, right @rodrigopinto ?

**Comment by rodrigopinto** [comment]
@thiagogabriel I suggest to not use pull, but raise a warning when the `push` command to break.
The HEAD must be clean to receive new push's, in other words, the merge should be made manually, and after that, git push will work fine. What do you think guys @maurogeorge @eduardofiorezi?

**Comment by eduardofiorezi** [comment]
I agree with @rodrigopinto, we will remove a magical command and let the power and sanity to the devs.

**Comment by maurogeorge** [comment]
I like the `git pull --rebase` :smile: 

@rodrigopinto I don't know where you find `--rebase` isn`t a good practice, could you please sent the link? About `-- rebase` or not I like this [post from Mislav](http://mislav.uniqpath.com/2013/02/merge-vs-rebase/).

If you do `git pull` you do the merge same way, if you have conflicts you need to solve this the same way.

**Comment by marcioj** [comment]
I agree with @maurogeorge. The only diference between --rebase and merge is that merge will create a new commit to join the commits. In all great OSS projects I see around, it's used the --rebase to keep the history clean.

**Comment by eduardofiorezi** [comment]
TL;DR https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/

Losing commits is much more dangerous than leaving the clean tree. It makes me give more value to merge.

**Comment by andreanastacio** [comment]
"Incorporates changes from a remote repository into the current branch. In its default mode, git pull is shorthand for git fetch followed by git merge FETCH_HEAD." http://git-scm.com/docs/git-pull

Using "git pull" solves the problem.

**Comment by andreanastacio** [comment]
#69

**Comment by maurogeorge** [comment]
@marano and @mateusg you guys have a optinion about that?

**Comment by marano** [comment]
Even though @rodrigopinto mentioned that `git rebase` is not a best practice, he was actually  suggesting the merge to be done manually. Not so much about choosing between a merge or a rebase. So I think #69 is pretty much a different matter. And I do agree with him, the integrate task should warn the user there are changes to be merged, and the user should do it manually before proceeding.

But I would like to say a few words about *how* to merge. The article [1] mentioned by @eduardofiorezi makes a very relevant statement:

"**Do you value more a clean, linear history? Or the traceability of your branches?** In the first case go for a **rebase** policy, in the later go for a **merge** one."

I do not completely agree with it, so I will try to explain *why* and *why* its still relevant. Merge and rebase can be used to achieve the same end goal, which is to share code changes with others. But they are not interchangeable because they are meant for different use cases, and the results are usually not the same.

Merge will incorporate the changes from another branch into the current branch. When there are conflicts, they are solved via a merge commit. Even when there are no conflicts, some people like to force an empty merge commit as a way to signal where the integration happened. And that's exactly what Github does when merging a pull request. That is important when it comes to keep track where branches got pulled to master. Merge is a way to make a statement on the history about when things happened regarding other branches. This is discussed on the article mentioned by @maurogeorge [2]. Also note, that this article heavily relies on feature branches  in order to make sense of both merge and rebase usages. And there we touch on the above statement - *"traceability of your branches"*. Note that the git merge `--no-ff` flag, which forces git to create an empty commit even if there are no conflicts, is not a git default, and is known to cause problems, the more aggravating of which is that it breaks git blame and bisect [3].

The git rebase command will get the changes from another branch, and applies the current branch changes on top of it. It forces you to solve each commit conflict, if there are any, as they are applied. Meaning that the conflict resolution will then be a part of that commit, as you went back in time, and redid your changes. So rebase will never need to create solving commits. And that's the *"more a clean, linear history"*. But the rebase conflict solving is harder to maneuver, and might lead to loss of data if not done correctly. Commands like `git checkout --theirs/--ours` and `git rerere` are tricks that need to be mastered in order to rebase bigger conflicts that extends through multiple commits.

With these differences of purpose in mind it is worth noting that the project workflow has great influence about the strategy a team will prefer. And in different situations, each will be more appropriated. The way I use merge and rebase looks like this:

git rebase:
- I am on master and want to get the changes from origin/master.
- I am on a branch and want to get the changes from origin/branch.
- I am on a (short lived) branch and want to get latest changes from master.

git merge:
- I am on master and want to get changes from a branch. I will always use `--ff-only`.
- I am on master and want to get changes from a branch, and I want to add a merge commit for the sake of history.
- I am on master and want to get huge set changes from a branch. Stuff like big refactoring, moving files around, library upgrades.

Note that, rebase and merge complement each other. Let's say I have changes on a branch that I want to merge with master. But my branch is outdated from master. So what I do is rebase the branch with master, solve whatever conflicts there are, and then proceed with to merge master with the branch. This allows me to avoid having merge commits with conflicts.

My opinion is that we should not proceed with #69, we should change the merge/rebase to an alert saying the branch cannot be pushed and needs to be merged first.

[1] https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/
[2] http://mislav.uniqpath.com/2013/02/merge-vs-rebase/
[3] https://sandofsky.com/blog/git-workflow.html

**Comment by rodrigopinto** [comment]
What Thiago wrote was exactly what I would like to explain, but he took a time that I didn't and explained really better than me. 

@marano I Just Called to Say I Love You. :heart: :heart: :heart:

**Comment by andreanastacio** [comment]
@marano :clap: :+1:
